Add instance norm

zsef123 opened the pull request:
# Description

Add `InstanceNorm` using[TensorRT InstanceNormalizationPlugin](https://github.com/NVIDIA/TensorRT/tree/master/plugin/instanceNormalizationPlugin) and implementation of BatchNorm

Fixes #210 

Also I checked #285 
But this PR not supported TensorRT 8++
And it's a little different from mine.

1. default beta value, BN and #285  set to 1 but I think that must be 0
2. type of eps, float is right type. And also my env, double occurs error.
3. I called TensorRT directly not modifying other context (follows `gelu`)

## Type of change

Please delete options that are not relevant and/or add your own.

- New feature (non-breaking change which adds functionality)
- This change requires a documentation update

# Checklist:

- [x] My code follows the style guidelines of this project (You can use the linters)
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas and hacks
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests to verify my fix or my feature
- [x] New and existing unit tests pass locally with my changes

zsef123 commented:
@narendasan passed DCO

narendasan commented:
@zsef123 Thanks for the contribution! Can you rebase your branch? Changes look mostly fine, once you rebase, I will run the tests and we should be good to go

zsef123 commented:
> @zsef123 Thanks for the contribution! Can you rebase your branch? Changes look mostly fine, once you rebase, I will run the tests and we should be good to go

Checked!!

narendasan commented on core/conversion/converters/impl/batch_norm.cpp:
```diff
@@ -10,61 +10,163 @@ namespace converters {
+
+              auto scales = args[1].unwrapToTensor(at::ones(shape[1], options)).cpu().contiguous();
+              auto bias = args[2].unwrapToTensor(at::zeros(shape[1], options)).cpu().contiguous();
+              LOG_DEBUG("Scales : " <<    );
```
I think you are missing the `scales`  here. Its causing a compilation error

narendasan commented:
I am getting test failures for Converters.ATenInstanceNormRunningStatsConvertsCorrectly where the TRT implementation returns NaN. Using TRT8

zsef123 commented:
@narendasan  Sorry for missed.

Several problems were mixed. All of them were test cases.
1. range of `running_mean` and `running_var`
In testcase, cause overflow when calculate `gamma` and `bias`.
So changed `randn` -> `ones`, `zeros`. It is default value of `instance_norm`

2. not clone parameters in test case
fixed

And changes
1. changed device
It is not bug, but I just follows `batch_norm`
Changed `cpu` to `cuda` when the `track_running_stats==True`

2. readable graph in test

P.S.
I had a hard time creating a new docker environment.
Any plans for a distribution with NGC 21.07?

narendasan commented:
@zsef123 check #578 for a 21.07 container. It uses a version of Pytorch slightly newer than 1.9 but other than that should work

narendasan reviewed (approved):
Cool, tests pass, LGTM
